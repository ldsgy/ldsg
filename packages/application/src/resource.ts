import { Resource } from "@ldsg/resource";
import express, { Express } from "express";
import { ApplicationSpecificResourceSettings } from "./types";

type CreateApp = () => Promise<Express>;

interface AddRoutesParams {
  app: Express;
}

type AddRoutes = (params: AddRoutesParams) => Promise<void>;

export class ApplicationResource extends Resource<ApplicationSpecificResourceSettings> {
  addRoutes: AddRoutes = async (params) => {
    const { app } = params;

    const {
      settings: { name },
      getFilteredResources,
    } = this;

    const getFilteredResourcesRes = getFilteredResources({});

    const { resources } = getFilteredResourcesRes;

    resources.forEach((resource) => {
      if ("addRoute" in resource && typeof resource.addRoute === "function") {
        resource?.addRoute();
      }
    });

    app.get("/", (req, res) => {
      res.send(`Hello, ${name ?? "World"}!`);
    });
  };

  createApp: CreateApp = async () => {
    const app = express();

    await this.addRoutes({
      app,
    });

    return app;
  };
}
