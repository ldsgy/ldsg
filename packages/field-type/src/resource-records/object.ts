import { ROOT_RESOURCE_ID } from "@ldsg/constants";
import {
  HandlerExtendedResourceSettings,
  HandlerSpecificResourceSettings,
} from "@ldsg/handler";
import { ResourceRecord } from "@ldsg/types";
import {
  FieldTypeBasePlatform,
  FieldTypeResourceHandler,
  FieldTypeSpecificResourceSettings,
} from "../types";

export const objectFieldTypeResourceRecord: ResourceRecord<
  HandlerExtendedResourceSettings<FieldTypeSpecificResourceSettings>
> = {
  id: "object-field-type",
  kind: "field_type",
  parentId: ROOT_RESOURCE_ID,
  settings: {
    title: "对象字段类型",
    description: "",
    fieldPropertiesSchema: {
      type: "object",
      properties: {},
    },
    handlerResourceId: "object-field-type-handler",
  },
};

const handler: FieldTypeResourceHandler = (params) => {
  const { fieldProperties, platform } = params;

  let res;

  switch (platform) {
    case FieldTypeBasePlatform.COMMON: {
      res = {
        ...fieldProperties,
        type: "object",
      };

      break;
    }

    default: {
      break;
    }
  }

  return res;
};

export const objectFieldTypeHandlerResourceRecord: ResourceRecord<HandlerSpecificResourceSettings> =
  {
    id: "object-field-type-handler",
    kind: "handler",
    parentId: ROOT_RESOURCE_ID,
    settings: {
      title: "对象字段类型处理程序",
      description: "",
      code: "",
      dependencies: [],
      handler,
    },
  };
