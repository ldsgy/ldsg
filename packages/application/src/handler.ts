import { Handler } from "@ldsg/handler";
import { Manifest } from "@ldsg/resource";
import { ApplicationResource } from "./resource";
import { ApplicationResourceSettings } from "./types";

export const handler: Handler<
  [Manifest.Resource<ApplicationResourceSettings>]
> = (params) => {
  return new ApplicationResource(params);
};
