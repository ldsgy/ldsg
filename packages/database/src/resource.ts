import { Resource, ResourceConstructorParams } from "@ldsg/resource";
import mongoose from "mongoose";
import { DatabaseSpecificResourceSettings } from "./types";

export class DatabaseResource extends Resource<DatabaseSpecificResourceSettings> {
  constructor(
    params: ResourceConstructorParams<DatabaseSpecificResourceSettings>
  ) {
    super(params);

    const { connect } = this;

    connect();
  }

  connect = () => {
    const {
      settings: { uri, connectOptions },
    } = this;

    return mongoose.connect(uri, connectOptions);
  };

  disconnect = () => {
    return mongoose.disconnect();
  };
}
