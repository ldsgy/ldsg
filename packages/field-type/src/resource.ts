import { Resource } from "@ldsg/resource";
import { FieldTypeSpecificResourceSettings } from "./types";

export class FieldTypeResource extends Resource<FieldTypeSpecificResourceSettings> {}
