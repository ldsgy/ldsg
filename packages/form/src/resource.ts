import { ObjectInfo, ObjectResource } from "@ldsg/object";
import { Resource } from "@ldsg/resource";
import { FormSpecificResourceSettings } from "./types";

interface GetFormInfoParams {
  /**
   * Platform
   * Such as mongoose\formily.
   */
  platform: string;
}

export interface FormInfo {
  /**
   * Form Name
   */
  name: string;

  /**
   * Input Object Info
   */
  inputObjectInfo: ObjectInfo;

  /**
   * Output Object Info
   */
  outputObjectInfo: ObjectInfo;
}

type GetFormInfo = (params: GetFormInfoParams) => FormInfo;

export class FormResource extends Resource<FormSpecificResourceSettings> {
  getFormInfo: GetFormInfo = (params) => {
    const { platform } = params;

    const {
      settings: { name },
      getResourcesFromSettings,
    } = this;

    const { inputObjectResource, outputObjectResource, workflowResource } =
      getResourcesFromSettings();

    const inputObjectInfo = (
      inputObjectResource as ObjectResource
    ).getObjectInfo({ platform });

    const outputObjectInfo = (
      outputObjectResource as ObjectResource
    ).getObjectInfo({ platform });

    const res: FormInfo = {
      name,
      inputObjectInfo,
      outputObjectInfo,
    };

    return res;
  };
}
