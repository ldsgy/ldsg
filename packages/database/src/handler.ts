import { Handler } from "@ldsg/handler";
import {
  InstantiateResourceParams,
  InstantiateResourceRes,
} from "@ldsg/resource";
import { instantiateResource } from "./instantiate-resource";
import { DatabaseResource } from "./resource";
import { DatabaseSpecificResourceSettings } from "./types";

export const handler: Handler<
  [InstantiateResourceParams<DatabaseSpecificResourceSettings>],
  InstantiateResourceRes<DatabaseResource>
> = instantiateResource;
