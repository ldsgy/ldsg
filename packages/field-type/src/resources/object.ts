import { ROOT_RESOURCE_ID } from "@ldsg/constants";
import { FieldTypeResource } from "@ldsg/field-type";
import { HandlerResource } from "@ldsg/handler";
import { FieldTypeBasePlatform, FieldTypeResourceHandler } from "../types";

export const objectFieldType = new FieldTypeResource({
  id: "object-field-type",
  kind: "field_type",
  parentId: ROOT_RESOURCE_ID,
  settings: {
    title: "对象字段类型",
    description: "",
    fieldPropertiesSchema: {
      type: "object",
      properties: {},
    },
    handlerResourceId: "object-field-type-handler",
  },
});

const handler: FieldTypeResourceHandler = (params) => {
  const { fieldProperties, platform } = params;

  let res;

  switch (platform) {
    case FieldTypeBasePlatform.COMMON: {
      res = {
        ...fieldProperties,
        type: "object",
      };

      break;
    }

    default: {
      break;
    }
  }

  return res;
};

export const objectFieldTypeHandler = new HandlerResource({
  id: "object-field-type-handler",
  kind: "handler",
  parentId: ROOT_RESOURCE_ID,
  settings: {
    title: "对象字段类型处理程序",
    description: "",
    code: "",
    dependencies: [],
  },
  handler,
});
