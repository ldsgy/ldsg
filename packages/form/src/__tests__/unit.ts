import { FieldTypeResource } from "@ldsg/field-type";
import { Handler, HandlerResource } from "@ldsg/handler";
import { ObjectResource } from "@ldsg/object";
import { ObjectFieldResource } from "@ldsg/object-field";
import { WorkflowResource } from "@ldsg/workflow";
import { FormResource } from "../resource";

const handler: Handler<
  [
    {
      /**
       * Field Properties
       */
      fieldProperties?: any;
      /**
       * Platform
       * Such as mongoose\formily.
       */
      platform: string;
    }
  ],
  {}
> = (params) => {
  const { fieldProperties, platform } = params;

  const { max } = fieldProperties;

  let res;

  switch (platform) {
    case "mongoose": {
      res = {
        type: "String",
        ...(max
          ? {
              maxLength: max,
            }
          : {}),
      };

      break;
    }

    default: {
      res = {};

      break;
    }
  }

  return res;
};

test("unit", () => {
  new HandlerResource({
    id: "test-handler",
    kind: "handler",
    parentId: "root",
    settings: {
      title: "测试处理程序",
      description: "",
      code: "",
      dependencies: [],
      handler,
    },
  });

  new FieldTypeResource({
    id: "test-field-type",
    kind: "field_type",
    parentId: "root",
    settings: {
      title: "文本",
      description: "可用来存储各种文本",
      fieldPropertiesSchema: {
        type: "object",
        properties: {
          max: {
            type: "integer",
            title: "配置最长字符",
            description: "长度不可以超过此值",
          },
        },
      },
      handlerResourceId: "test-handler",
    },
  });

  new ObjectResource({
    id: "test-object-1",
    kind: "object",
    parentId: "root",
    settings: {
      title: "测试对象1",
      description: "",
      name: "test-object-1",
    },
  });

  new ObjectFieldResource({
    id: "test-object-1-field-1",
    kind: "object_field",
    parentId: "test-object-1",
    settings: {
      title: "测试对象1测试字段1",
      description: "",
      name: "test-1",
      fieldTypeResourceId: "test-field-type",
      properties: {
        max: "10",
      },
    },
  });

  new ObjectFieldResource({
    id: "test-object-1-field-2",
    kind: "object_field",
    parentId: "test-object-1",
    settings: {
      title: "测试对象1测试字段2",
      description: "",
      name: "test-2",
      fieldTypeResourceId: "test-field-type",
      properties: {
        max: "10",
      },
    },
  });

  new ObjectFieldResource({
    id: "test-object-1-field-3",
    kind: "object_field",
    parentId: "test-object-1",
    settings: {
      title: "测试对象1测试字段3",
      description: "",
      name: "test-3",
      fieldTypeResourceId: "test-field-type",
      properties: {
        max: "10",
      },
    },
  });

  new ObjectResource({
    id: "test-object-2",
    kind: "object",
    parentId: "root",
    settings: {
      title: "测试对象2",
      description: "",
      name: "test-object-2",
    },
  });

  new ObjectFieldResource({
    id: "test-object-2-field-1",
    kind: "object_field",
    parentId: "test-object-2",
    settings: {
      title: "测试对象2测试字段1",
      description: "",
      name: "test-1",
      fieldTypeResourceId: "test-field-type",
      properties: {
        max: "10",
      },
    },
  });

  new WorkflowResource({
    id: "test-workflow",
    kind: "workflow",
    parentId: "root",
    settings: {
      title: "测试工作流",
      description: "",
    },
  });

  const formResource = new FormResource({
    id: "test-form",
    kind: "form",
    parentId: "forms",
    settings: {
      title: "测试表单",
      description: "",
      name: "test-form",
      inputObjectResourceId: "test-object-1",
      outputObjectResourceId: "test-object-2",
      workflowResourceId: "test-workflow",
    },
  });

  const getFormInfoRes = formResource.getFormInfo({
    platform: "mongoose",
  });

  expect(getFormInfoRes).toMatchSnapshot();
});
