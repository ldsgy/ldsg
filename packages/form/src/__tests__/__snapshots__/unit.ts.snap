// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit 1`] = `
{
  "description": "",
  "id": "test-form",
  "inputObjectInfo": {
    "description": "",
    "fieldInfoList": [
      {
        "description": "",
        "id": "test-object-1-field-1",
        "name": "test-1",
        "title": "测试对象1测试字段1",
        "typeInfo": {
          "platformToTypeMap": {
            "COMMON": {
              "max": "10",
              "type": "string",
            },
            "GRAPHQL": undefined,
            "JSON": undefined,
            "MONGOOSE": undefined,
          },
        },
      },
      {
        "description": "",
        "id": "test-object-1-field-2",
        "name": "test-2",
        "title": "测试对象1测试字段2",
        "typeInfo": {
          "platformToTypeMap": {
            "COMMON": {
              "max": "10",
              "type": "string",
            },
            "GRAPHQL": undefined,
            "JSON": undefined,
            "MONGOOSE": undefined,
          },
        },
      },
      {
        "description": "",
        "id": "test-object-1-field-3",
        "name": "test-3",
        "title": "测试对象1测试字段3",
        "typeInfo": {
          "platformToTypeMap": {
            "COMMON": {
              "max": "10",
              "type": "string",
            },
            "GRAPHQL": undefined,
            "JSON": undefined,
            "MONGOOSE": undefined,
          },
        },
      },
    ],
    "name": "test-object-1",
    "title": "测试对象1",
  },
  "name": "test-form",
  "outputObjectInfo": {
    "description": "",
    "fieldInfoList": [
      {
        "description": "",
        "id": "test-object-2-field-1",
        "name": "test-1",
        "title": "测试对象2测试字段1",
        "typeInfo": {
          "platformToTypeMap": {
            "COMMON": {
              "max": "10",
              "type": "string",
            },
            "GRAPHQL": undefined,
            "JSON": undefined,
            "MONGOOSE": undefined,
          },
        },
      },
    ],
    "name": "test-object-2",
    "title": "测试对象2",
  },
  "title": "测试表单",
  "workflowInfo": {
    "execute": [Function],
  },
}
`;
